iOS: Provide correct animation function for hiding keyboard

rigdern opened the pull request:
Currently, `7 << 16` is used as the animation easing function for both the showing and hiding of the keyboard.

Instead, `7 << 16` should be used for showing the keyboard and `6 << 16` for hiding the keyboard.

For details, see: http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve/19439283#19439283

**Test plan (required)**

Testing animating the position of a view to coordinate with the keyboard hiding and showing and it looks good.

Adam Comella
Microsoft Corp.

satya164 commented:
cc @grabbou

ide commented on React/Base/RCTKeyboardObserver.m:
```diff
@@ -94,11 +94,12 @@ - (void)EVENT:(NSNotification *)notification \
   CGRect endFrame = [userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
   NSTimeInterval duration = [userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
   UIViewAnimationCurve curve = [userInfo[UIKeyboardAnimationCurveUserInfoKey] integerValue];
+  bool isShowing = beginFrame.origin.y > endFrame.origin.y;
```
Rename this (and the param in the function above) to `willShowKeyboard`

ide commented on React/Modules/RCTUIManager.m:
```diff
@@ -78,8 +78,14 @@ static UIViewAnimationOptions UIViewAnimationOptionsFromRCTAnimationType(RCTAnim
       return UIViewAnimationOptionCurveEaseInOut;
     case RCTAnimationTypeKeyboard:
-      // http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve
+      // http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve/19439283#19439283
```
Can you add a log message (RCTLogWarn) when "keyboard" is specified so that we can deprecate and later remove this value, unless you can think of a reason to keep it around?

rigdern commented:
@ide Thanks for the feedback. I pushed the changes you requested.

nicklockwood commented on React/Base/RCTConvert.m:
```diff
@@ -842,6 +842,8 @@ + (NSPropertyList)NSPropertyList:(id)json
   @"easeOut": @(RCTAnimationTypeEaseOut),
   @"easeInEaseOut": @(RCTAnimationTypeEaseInEaseOut),
   @"keyboard": @(RCTAnimationTypeKeyboard),
```
"keyboard"  should just be an alias for RCTAnimationTypeKeyboardShow

nicklockwood commented on Libraries/LayoutAnimation/LayoutAnimation.js:
```diff
@@ -24,6 +24,8 @@ var TypesEnum = {
   easeIn: true,
   easeOut: true,
   keyboard: true,
```
Please document this as being deprecated

nicklockwood commented on React/Views/RCTAnimationType.h:
```diff
@@ -16,4 +16,6 @@ typedef NS_ENUM(NSInteger, RCTAnimationType) {
   RCTAnimationTypeEaseOut,
   RCTAnimationTypeEaseInEaseOut,
   RCTAnimationTypeKeyboard,
```
Remove this - there are no call sites for it on the native side except for the one's you've updated in this diff.

nicklockwood commented on React/Modules/RCTUIManager.m:
```diff
@@ -78,8 +78,15 @@ static UIViewAnimationOptions UIViewAnimationOptionsFromRCTAnimationType(RCTAnim
     case RCTAnimationTypeKeyboard:
-      // http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve
+      // http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve/19439283#19439283
+      RCTLogWarn(@"The animation type RCTAnimationTypeKeyboard is deprecated. Instead use RCTAnimationTypeKeyboardShow or RCTAnimationTypeKeyboardHide.");
```
Good idea, but can you move this deprecation warning to the JS side instead, that way the legacy support code can be kept only on JS side, making it easier to remove later.

nicklockwood commented:
I'm actually wondering, given how this constant is to be used, whether we can implement this in a more robust/elegant way.

Instead of hard coding the value of keyboard, what if we record it from the keyboardWillShow/Hide event?

The only time you'd be using the "keyboard" value for a layout animation is when the keyboard has just started to show or hide, so in theory the keyboard notification should always have fired just before we trigger the animation, and so we can record and use the correct value without needing two separate constants.

facebook-github-bot-9 commented:
@rigdern updated the pull request.

rigdern commented:
@nicklockwood I implemented your suggestion of remembering state from the keyboard event so that we only need one keyboard-related animation constant.

I also changed the approach a bit -- instead of hardcoding the animation curve, I now read it off of the keyboard event. According to the comments on [this stackoverflow post](http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve/19439283#19439283), the keyboard event has been providing accurate curve information since iOS 7.1.

rigdern commented:
@nicklockwood Did you have a chance to look at the revisions I made based on your feedback?

Adam Comella
Microsoft Corp.

nicklockwood commented on React/Modules/RCTUIManager.m:
```diff
@@ -64,6 +64,8 @@ @interface RCTAnimation : NSObject
 
 @end
 
+static int _currentKeyboardAnimationCurve;
```
Can you make this `UIViewAnimationCurve` instead of `int`? I know that the value it's set to doesn't actually appear in the `UIViewAnimationCurve` enum, but it's intended to be passed to a method parameter of that type so arguably it's a private member of that enum.

nicklockwood commented on React/Modules/RCTUIManager.m:
```diff
@@ -79,13 +81,30 @@ static UIViewAnimationOptions UIViewAnimationOptionsFromRCTAnimationType(RCTAnim
++ (void)keyboardWillChangeFrame:(NSNotification *)notification
+{
+  NSDictionary *userInfo = notification.userInfo;
+  _currentKeyboardAnimationCurve = [userInfo[UIKeyboardAnimationCurveUserInfoKey] intValue];
```
This should be `integerValue` rather than `intValue`, assuming you make the change above.

nicklockwood commented on React/Modules/RCTUIManager.m:
```diff
@@ -79,13 +81,30 @@ static UIViewAnimationOptions UIViewAnimationOptionsFromRCTAnimationType(RCTAnim
   }
 }
 
++ (void)initializeStatics
```
This could just be done in the `+(void)initialize` method, then you wouldn't need to explicitly call it.

rigdern replied:
I can't use `+initialize` here. I'll add a comment describing why. Here's the reasoning: `+initialize` runs immediately before the first message is sent to the class which may be too late for us. By this time, we may have missed some `UIKeyboardWillChangeFrameNotification`s.

I might be able to use `+load` but there are some [caveats when using `+load`](https://www.mikeash.com/pyblog/friday-qa-2009-05-22-objective-c-class-loading-and-initialization.html) because it is executed so early so I feel more comfortable with a custom initialization function.

nicklockwood commented:
@rigdern awesome - thanks for doing that. A few minor nits, but otherwise this should be good to go.

rigdern commented:
@nicklockwood I addressed your feedback.

Adam Comella
Microsoft Corp.

nicklockwood commented:
@facebook-github-bot shipit

facebook-github-bot-5 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1005977826149758/int_phab) to review.
